src/elements/content-sharing/ContentSharing.js

```diff
@@ -0,0 +1,175 @@
+    const [item, setItem] = useState<?itemFlowType>(null);
+    const [sharedLink, setSharedLink] = useState<?ContentSharingSharedLinkType>(null);
+    const [currentUserID, setCurrentUserID] = useState<?string>(null);
+    const [errorExists, setErrorExists] = useState<boolean>(false);
+
+    const api = new API({
+        apiHost,
+        clientName: CLIENT_NAME_CONTENT_SHARING,
+        id: `${itemType}_usm`,
+        token,
+    });
+
+    const resetState = useCallback(() => {
+        setItem(null);
+        setSharedLink(null);
+        setCurrentUserID(null);
+    }, [setItem, setSharedLink, setCurrentUserID]);
+
+    const getError = useCallback(() => {
+        setErrorExists(true);
+    }, [setErrorExists]);
+
+    const getUserSuccess = useCallback(
+        userData => {
+            const { id, userEnterpriseData } = normalizeUserResponse(userData);
+            setSharedLink({ ...sharedLink, ...userEnterpriseData });
+            setCurrentUserID(id);
+            setErrorExists(false);
+        },
+        [sharedLink],
+    );
+    const getItemSuccess = (itemData: ContentSharingItemAPIResponse) => {
+        const { item: itemFromAPI, sharedLink: sharedLinkFromAPI } = normalizeItemResponse(itemData);
+        setItem(itemFromAPI);
+        setSharedLink(sharedLinkFromAPI);
+        setErrorExists(false);
+    };
+
+    useEffect(() => {
+        resetState();
+    }, [resetState, token, itemID, itemType]);
+
+    useEffect(() => {
+        const getItem = async () => {
+            if (itemType === TYPE_FILE) {
+                await api.getFileAPI().getFile(itemID, getItemSuccess, getError, {
+                    fields: [
+                        FIELD_ALLOWED_INVITEE_ROLES,
+                        FIELD_DESCRIPTION,
+                        FIELD_EXTENSION,
+                        FIELD_ID,
+                        FIELD_NAME,
+                        FIELD_PERMISSIONS,
+                        FIELD_SHARED_LINK,
+                        FIELD_SHARED_LINK_FEATURES,
+                        FIELD_ITEM_TYPE,
+                    ],
+                });
+            }
+
+            if (itemType === TYPE_FOLDER) {
+                await api.getFolderAPI().getFolderFields(itemID, getItemSuccess, getError, {
+                    fields: [
+                        FIELD_ALLOWED_INVITEE_ROLES,
+                        FIELD_DESCRIPTION,
+                        FIELD_EXTENSION,
+                        FIELD_ID,
+                        FIELD_NAME,
+                        FIELD_PERMISSIONS,
+                        FIELD_SHARED_LINK,
+                        FIELD_SHARED_LINK_FEATURES,
+                        FIELD_ITEM_TYPE,
+                    ],
+                });
+            }
+        };
+
+        if (!item && !sharedLink) {
+            getItem();
+        }
+    }, [api, getError, item, itemID, itemType, sharedLink, currentUserID]);
```

Comment:
This may be a silly question. item, itemID, itemType, and sharedLink are all related. I am not fully informed on how `useEffect` works, but curious if it will be triggered unnecessarily, given these somewhat derived variables. 

Example: item has `.id` and `.type`, and `.sharedLink (`BoxItem`). `useEffect` presumably responds to these values changing. could we use fewer of these without triggering the correct number of side-effects? same for the others with this subset

Reply:
By default, React runs each `useEffect` function whenever any of the values in the component scope changes; specifying this array allows React to skip the effect if none of the values in the array has changed. So, this is actually an optimization! The array has to include all values that the effect uses, so that the references to the values remain current (also, I get a lint error if I try to remove any of them)